### PR TITLE
Add asio coroutine support

### PR DESCRIPTION
**Pull request by yurp**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -52,6 +49,7 @@ target_include_directories(asio-dbus INTERFACE
                            $<INSTALL_INTERFACE:${CMAKE_INSTALL_PREFIX}/include>
 			   ${DBUS_INCLUDE_DIRS})
 
+target_compile_features(asio-dbus INTERFACE cxx_std_20)
```
Comment (zagrichanskiy) [inline CMakeLists.txt]: library is headeronly, so this should be enabled only for tests executable

File: include/dbus/connection_service.hpp [hunk 2]
```diff
@@ -77,18 +77,18 @@ class connection_service
...
-    return init.result.get();
+                 ASIO_MOVE_ARG(MessageHandler) h, int timeout_ms) {
+
+    auto i = asio::bind_executor(this->get_io_context().get_executor(), [this, &impl, &m, timeout_ms](auto&& handler)
+    {
+      // begin asynchronous operation
+      impl.start(this->get_io_context());
+
+      using handler_type = std::decay_t<decltype(handler)>;
+      detail::async_send_op<handler_type>(this->get_io_context(), std::forward<handler_type>(handler)) (impl, m, timeout_ms);
```
Comment (zagrichanskiy) [inline include/dbus/connection_service.hpp]: std::forward<handler_type>(handler))
it makes no sense to forward already decayed type

File: include/dbus/detail/async_send_op.hpp [hunk 3]
```diff
@@ -25,15 +25,15 @@ struct async_send_op {
...
+  async_send_op(asio::io_context& io, MessageHandler&& handler);
   static void callback(DBusPendingCall* p, void* userdata);  // for C API
   void operator()(impl::connection& c, message& m, int timeout_ms);  // initiate operation
   void operator()();  // bound completion handler form
 };
 
 template <typename MessageHandler>
 async_send_op<MessageHandler>::async_send_op(asio::io_context& io,
-                                             MessageHandler& handler)
+                                             MessageHandler&& handler)
```
Comment (zagrichanskiy) [inline include/dbus/detail/async_send_op.hpp]: MessageHandler is a class template parameter, so in scope of the constructor it's already a concrete type
MessageHandler&& here is an rvalue reference, not forwarding reference

Comment (zagrichanskiy) [reply include/dbus/detail/async_send_op.hpp]: I think it was correct initially

File: include/dbus/connection_service.hpp [hunk 4]
```diff
@@ -77,18 +77,18 @@ class connection_service
...
+    });
+
+    return asio::async_initiate<MessageHandler, void(asio::error_code, message)>(std::move(i), h);
```
Comment (zagrichanskiy) [inline include/dbus/connection_service.hpp]: to move from lambda you should declare it mutable (however there is nothing to move from it really)

File: include/dbus/detail/queue.hpp [hunk 5]
```diff
@@ -68,32 +68,27 @@ class queue {
...
+        messages.pop_front();
+
+        asio::post(io, [h = std::forward<decltype(handler)>(handler), m = std::move(m)]
+        {
+          h({}, std::move(m));
+        });
+      }
+    });
+
+    return asio::async_initiate<MessageHandler, void(asio::error_code, message_type)>(std::move(i), h);
```
Comment (zagrichanskiy) [inline include/dbus/detail/queue.hpp]: mutable if std::move, or remove std::move

File: test/avahi.cpp [hunk 6]
```diff
@@ -17,6 +18,49 @@
 
 using namespace std::literals;
 
+TEST(AvahiTest, GetHostNameCoAwait) {
+    asio::io_context io;
+
+    asio::co_spawn(io,
+        [&io]() ->asio::awaitable<void>
```
Comment (zagrichanskiy) [inline test/avahi.cpp]: auto ex = co_await asio::this_coro::executor
instead of &io

File: test/avahi.cpp [hunk 7]
```diff
@@ -17,6 +18,49 @@
...
+            gethostname(c, 1024);
+            hostname = c;
+
+            r.unpack(avahi_hostname);
+
+            // Get only the host name, not the fqdn
+            auto unix_hostname = hostname.substr(0, hostname.find("."));
+            EXPECT_EQ(unix_hostname, avahi_hostname);
+
+            io.stop();
```
Comment (zagrichanskiy) [inline test/avahi.cpp]: there is no asio::io_context::work created anywhere, stop is not needed

File: test/avahi.cpp [hunk 8]
```diff
@@ -17,6 +18,49 @@
...
+        },
+        [&io](std::exception_ptr )
+        {
+            io.stop();
```
Comment (zagrichanskiy) [inline test/avahi.cpp]: just rethrow for the test to exit with a failure

File: test/avahi.cpp [hunk 9]
```diff
@@ -17,6 +18,49 @@
...
+        });
+
+
+  asio::steady_timer t(io, 10s);
+
+  t.async_wait([&](const asio::error_code& /*e*/) {
```
Comment (zagrichanskiy) [inline test/avahi.cpp]: remove timer
just invoke io.run_for(10s) instead

File: test/connection.cpp [hunk 10]
```diff
@@ -67,7 +67,7 @@ TEST(ConnectionTest, CallWithExpiredTimeoutAndReplyCallback) {
     bool test_result = false;
     dbus::message m = dbus::message::new_call({srv_name, "/", srv_name, srv_method});
     m.pack(42);
-    client.async_send(m, [&io, &test_result] (asio::error_code ec, dbus::message& r)
+    client.async_send(m, [&io, &test_result] (asio::error_code ec, dbus::message r)
```
Comment (zagrichanskiy) [inline test/connection.cpp:70]: please ensure that library moves a message, otherwise we may perform unneeded copy

**Review by zagrichanskiy — CHANGES_REQUESTED** [review]

File: test/avahi.cpp (hunk 8 above)
Comment (yurp) [reply test/avahi.cpp]: agree with rethrow

File: include/dbus/connection_service.hpp (hunk 4 above)
Comment (yurp) [reply include/dbus/connection_service.hpp]: agree

File: CMakeLists.txt (hunk 1 above)
Comment (yurp) [reply CMakeLists.txt]: ok

File: include/dbus/detail/queue.hpp (hunk 5 above)
Comment (yurp) [reply include/dbus/detail/queue.hpp]: ok

File: test/connection.cpp (hunk 10 above)
Comment (pernatiy) [reply test/connection.cpp:70]: dbus::message is a shared pointer holder

File: test/avahi.cpp (hunk 9 above)
Comment (yurp) [reply test/avahi.cpp]: ok

File: test/avahi.cpp (hunk 6 above)
Comment (yurp) [reply test/avahi.cpp]: done

File: test/avahi.cpp (hunk 7 above)
Comment (yurp) [reply test/avahi.cpp]: done

File: test/connection.cpp (hunk 10 above)
Comment (yurp) [reply test/connection.cpp:70]: looks risky, let's copy so far

File: include/dbus/detail/async_send_op.hpp (hunk 3 above)
Comment (yurp) [reply include/dbus/detail/async_send_op.hpp]: agree

File: include/dbus/connection_service.hpp (hunk 2 above)
Comment (yurp) [reply include/dbus/connection_service.hpp]: thanks

**Review by pernatiy — APPROVED** [review]

**Pull request merged**